### PR TITLE
CancellationToken not cancelling

### LINKED ISSUE #236: CancellationTokenSource not work
I need to stop the container after it runs longer than the time I give, then I try to use CancellationTokenSource, here is my code:

``` csharp
public static async Task DockerCancell()
{
	var cert = new CertificateCredentials(new X509Certificate2(CertPath, CertPassword));
	var configuration = new DockerClientConfiguration(new Uri("https://docker-1:2376"), cert);
	var client = configuration.CreateClient();
	var createResult = await client.Containers.CreateContainerAsync(new CreateContainerParameters()
	{
		Image = Image,
		Name = Guid.NewGuid().ToString(),
		WorkingDir = "/workdir",
		HostConfig = new HostConfig() { AutoRemove = true },
		Cmd = new[] { "sleep", "3" }
	});
	await client.Containers.StartContainerAsync(createResult.ID, new ContainerStartParameters());
	var token = new CancellationTokenSource();
	token.CancelAfter(1000);
	await client.Containers.WaitContainerAsync(createResult.ID, token.Token);
	Console.WriteLine("shouldn't be here");
}
```

It always output "shouldn't be here", looks like cancellation token didn't work for this api.
Further more, not only `CancelAfter`, `Cancel` is not work too:

``` csharp
public static async Task DockerCancell()
{
	var cert = new CertificateCredentials(new X509Certificate2(CertPath, CertPassword));
	var configuration = new DockerClientConfiguration(new Uri("https://docker-1:2376"), cert);
	var client = configuration.CreateClient();
	var createResult = await client.Containers.CreateContainerAsync(new CreateContainerParameters()
	{
		Image = Image,
		Name = Guid.NewGuid().ToString(),
		WorkingDir = "/workdir",
		HostConfig = new HostConfig() { AutoRemove = true },
		Cmd = new[] { "sleep", "3" }
	});
	await client.Containers.StartContainerAsync(createResult.ID, new ContainerStartParameters());
	var token = new CancellationTokenSource();
	// token.CancelAfter(1000);
	var task = client.Containers.WaitContainerAsync(createResult.ID, token.Token);
	await Task.Delay(1);
	token.Cancel();
	await task;
	Console.WriteLine("shouldn't be here");
}
```

### LINKED ISSUE #375: Response stream corruption when a request is cancelled
**Output of `dotnet --info`:**

```
.NET Core SDK (reflecting any global.json):
 Version:   2.1.503
 Commit:    4c506e0f35

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.503\

Host (useful for support):
  Version: 2.1.7
  Commit:  cca5d72d48

.NET Core SDKs installed:
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.402 [C:\Program Files\dotnet\sdk]
  2.1.503 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
```

**What version of Docker.DotNet?:**

```
// Docker.DotNet, Version=3.125.2.0, Culture=neutral, PublicKeyToken=null
// Global type: <Module>
// Architecture: AnyCPU (64-bit preferred)
// Runtime: .NET 4.0
```

**Steps to reproduce the issue:**

```
using System;
using System.Diagnostics;
using System.Threading;
using Docker.DotNet.Models;
using Newtonsoft.Json;

namespace ConsoleApp13
{
    class DockerSystemEventLogger : IProgress<JSONMessage>
    {
        /// <inheritdoc />
        public void Report(JSONMessage value)
        {
            Console.WriteLine($"DockerSystemEvent: {JsonConvert.SerializeObject(value)}");
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            var rand = new Random();
            var sw = new Stopwatch();

            using (var client = new Docker.DotNet.DockerClientConfiguration(new Uri("npipe://./pipe/docker_engine")).CreateClient())
            {
                for (int i = 0; i < 10000; ++i)
                {
                    try
                    {
                        // (1) Create monitor task
                        var cts = new CancellationTokenSource();
                        var monitorTask = client.System.MonitorEventsAsync(new ContainerEventsParameters(), new DockerSystemEventLogger(), cts.Token);

                        // (2) Wait for some time to make sure we get into blocking IO call
                        Thread.Sleep(100);

                        // (3) Invoke another request that will attempt to grab the same buffer
                        var listImagesTask1 = client.Images.TagImageAsync("microsoft/windowsservercore:ltsc2016", new ImageTagParameters { RepositoryName = "microsoft/windowsservercore", Tag = $"ltsc2016c", Force = true });

                        // (4) Wait for a short bit again and cancel the monitor task - if we get lucky, we the list images call will grab the same buffer while
                        sw.Restart();
                        var iterations = rand.Next(15000000);

                        for (int j = 0; j < iterations; j++)
                        {
                            // noop
                        }
                        Console.WriteLine($"Waited for {sw.Elapsed.TotalMilliseconds} ms");

                        cts.Cancel();

                        listImagesTask1.GetAwaiter().GetResult();
                        client.Images.TagImageAsync("microsoft/windowsservercore:ltsc2016", new ImageTagParameters { RepositoryName = "microsoft/windowsservercore", Tag = $"ltsc2016c", Force = true }).GetAwaiter().GetResult();

                        monitorTask.GetAwaiter().GetResult();

                        Console.WriteLine(i);
                    }
                    catch (Exception ex)
                    {
                        try
                        {
                            Console.ForegroundColor = ConsoleColor.Red;
                            Console.WriteLine($"{i} failed: {ex}");
                            ////break;
                        }
                        finally
                        {
                            Console.ResetColor();
                        }
                    }
                }
            }

            Console.ReadLine();
        }
    }
}
```

**What actually happened?:**

DockerClient.System.MonitorEventsAsync throws ArgumentException or IOException on cancellation. This results in one of three situations that we observed:

1. "System.ArgumentException: Stream was not readable."
2. "System.IO.IOException: Invalid chunk header: HTTP/1.1 201 Created".
3. "System.IO.IOException: Unexpected end of stream"

image

image

**What did you expect to happen?:**

OperationCancelledException to be thrown instead.

**Additional information:**

The problem is with this code:

https://github.com/Microsoft/Docker.DotNet/blob/master/src/Docker.DotNet/Endpoints/StreamUtil.cs

```
internal static async Task MonitorStreamForMessagesAsync<T>(Task<Stream> streamTask, DockerClient client, CancellationToken cancel, IProgress<T> progress)
{
    using (var stream = await streamTask)
    {
        // ReadLineAsync must be cancelled by closing the whole stream.
        using (cancel.Register(() => stream.Dispose()))
        {
            using (var reader = new StreamReader(stream, new UTF8Encoding(false)))
            {
                string line;
                try
                {
                    while ((line = await reader.ReadLineAsync()) != null)
                    {
                        var prog = client.JsonSerializer.DeserializeObject<T>(line);
                        if (prog == null) continue;

                        progress.Report(prog);
                    }
                }
                catch (ObjectDisposedException)
                {
                    // The subsequent call to reader.ReadLineAsync() after cancellation
                    // will fail because we disposed the stream. Just ignore here.
                }
            }
        }
    }
}
```

Basically on cancellation, cancel.Register disposes of the underlying stream while concurrently the request is being executed - can be in any stage...writing request data, reading response, etc. This is not thread safe, and all the three exceptions stem from it.

(1) Happens when the cancellation follows almost immediately after call to DockerClient.System.MonitorEventsAsync. 

If the cancellation token is signaled before the ChunkedReadStream is passed into the stream reader, it will be disposed, which will result in "System.ArgumentException: Stream was not readable." thrown out of StreamReader ctor.

(2) Happens due to corruption of the buffer owned by the BufferedReadStream. 

On cancellation, the underlying streams are disposed, including the BufferedReadStream. This will cause its buffer (currently owned by the BufferedReadStream) to be released back to the ArrayPool to be reused by another request. However, the read task itself is not cancelled, so now we have two active consumers of the same buffer. The next read on the "cancelled" monitor request results in invalid data - since the buffer was in the meantime populated by headers of another response.

Here's an example of one such buffer from a dump:

```
HTTP/1.1 201 Created
Api-Version: 1.39
Content-Length: 0
Date: Thu, 28 Feb 2019 15:35:21 GMT
Docker-Experimental: false
Ostype: windows
Server: Docker/18.09.0 (windows)
Connection: close

7394f62a4ca490fbce5","Attributes":{"name":"root/image:42"}},"scope":"local","time":1551368121,"timeNano":1551368121566912100}

}
```

This was a buffer that was being used and parsed by a ChunkedReadStream that was in the middle of monitoring system events. Hence "System.IO.IOException: Invalid chunk header: HTTP/1.1 201 Created"

(3) A similar problem as (2) with buffer data corruption, just swapped roles. 

1. Another request loads data into its buffer, 
2. System event monitor overwrites it with its next chunk that arrived on its pipe.
3. Then another request instead of reading an empty string to terminate and not ask for more data reads the system event monitor's chunk size. Assuming there's more data, it asks for the next line, but the underlying pipe stream is already broken since there should be no more data.

Hope this helps.

### LINKED ISSUE #416: Docker client 'defaultTimeout' not working
Using the following code, I would expect an exception to be raised in 5 seconds. Instead I am getting a OperationCanceledException at 20 seconds.

If I remove the defaultTimeout argument from the DockerClientConfiguration constructor, I no longer get an OperationCanceledException, but I do get a HttpRequestException after 20 seconds.

No matter what I change, an exception is always raised at 20 seconds. This is the code I am using (the IP address will be set to a host that is offline/unavailable).

Using Docker.DotNet 3.125.2 / .Net Core 2.2.

```
var client = new DockerClientConfiguration( new Uri( $"http://X.X.X.X:2375" ), defaultTimeout: TimeSpan.FromSeconds( 5 ) ).CreateClient();

var containers = await client.Containers.ListContainersAsync( new ContainersListParameters
{
	All = true
},
default );
```

### LINKED ISSUE #434: Events monitoring stream does not seems to always include CR or LF character at the end of every messages
**Output of `dotnet --info`:**
```
.NET Core SDK (reflecting any global.json):
 Version:   3.1.101
 Commit:    b377529961

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18362
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.101\

Host (useful for support):
  Version: 3.1.1
  Commit:  a1388f194c
```

**What version of Docker.DotNet?:**
```
3.125.2
```

**Docker Version:**
```
Client: Docker Engine - Community
 Version:           19.03.5
 API version:       1.40
 Go version:        go1.12.12
 Git commit:        633a0ea
 Built:             Wed Nov 13 07:22:37 2019
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.5
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.12
  Git commit:       633a0ea
  Built:            Wed Nov 13 07:29:19 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Docker Info:**
```
Client:                                                                                                   
 Debug Mode: false                                                                                        
                                                                                                          
Server:                                                                                                   
 Containers: 1                                                                                            
  Running: 1                                                                                              
  Paused: 0                                                                                               
  Stopped: 0                                                                                              
 Images: 10                                                                                               
 Server Version: 19.03.5                                                                                  
 Storage Driver: overlay2                                                                                 
  Backing Filesystem: extfs                                                                               
  Supports d_type: true                                                                                   
  Native Overlay Diff: true                                                                               
 Logging Driver: json-file                                                                                
 Cgroup Driver: cgroupfs                                                                                  
 Plugins:                                                                                                 
  Volume: local                                                                                           
  Network: bridge host ipvlan macvlan null overlay                                                        
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog                     
 Swarm: inactive                                                                                          
 Runtimes: runc                                                                                           
 Default Runtime: runc                                                                                    
 Init Binary: docker-init                                                                                 
 containerd version: b34a5c8af56e510852c35414db4c1f4fa6172339                                             
 runc version: 3e425f80a8c931f88e6d94a8c831b9d5aa481657                                                   
 init version: fec3683                                                                                    
 Security Options:                                                                                        
  seccomp                                                                                                 
   Profile: default                                                                                       
 Kernel Version: 4.19.76-linuxkit                                                                         
 Operating System: Docker Desktop                                                                         
 OSType: linux                                                                                            
 Architecture: x86_64                                                                                     
 CPUs: 2                                                                                                  
 Total Memory: 1.943GiB                                                                                   
 Name: docker-desktop                                                                                     
 ID: XESR:K73P:U465:PX4V:WDXD:HT4Y:H7K3:W7ED:7R2Y:KAEB:VF3G:WM7D                                          
 Docker Root Dir: /var/lib/docker                                                                         
 Debug Mode: true                                                                                         
  File Descriptors: 49                                                                                    
  Goroutines: 67                                                                                          
  System Time: 2020-01-31T17:25:24.631755878Z                                                             
  EventsListeners: 6                                                                                      
 Registry: https://index.docker.io/v1/                                                                    
 Labels:                                                                                                  
 Experimental: false                                                                                      
 Insecure Registries:                                                                                     
  127.0.0.0/8                                                                                             
 Live Restore Enabled: false                                                                              
 Product License: Community Engine                                                                        
```

**Steps to reproduce the issue:**
1.  Create DockerClient to local engine (npipe://./pipe/docker_engine) 
2.  Monitor system events ()
```
m_dockerClient.System.MonitorEventsAsync(
    new ContainerEventsParameters(),
    new Progress<Message>(m => Console.WriteLine(JsonConvert.SerializeObject(m, Newtonsoft.Json.Formatting.Indented))));
```
3. Using Docker CLI, monitor events (docker events)
4. Using the docker CLI, start and stops containers
5. Messages are printed on the Docker CLI console, but not on my app console (see 2.)

**What actually happened?:**
The HTTP request to get the event stream is accepted (200 - OK) and the stream is correctly opened, but there is no CR or LF character to delimit the messages received. Therefore, the stream reader gets stuck (since it is doing ReadLineAsync(...) calls to read messages and parse them) 

**What did you expect to happen?:**
Find a way to extract messages from the event stream without requiring CR or LF charcaters to detect a complete message.

A solution could be to directly parse the stream to extratc JSON objects/messages.

```
internal static async Task MonitorStreamForMessagesAsync<T>(Task<Stream> streamTask, DockerClient client, CancellationToken cancel, IProgress<T> progress)
{
    using (var stream = await streamTask)
    {
        // ReadLineAsync must be cancelled by closing the whole stream.
        using (cancel.Register(() => stream.Dispose()))
        {
            using (var reader = new StreamReader(stream, new UTF8Encoding(false))) 
            using (var jsonReader = new JsonTextReader(reader) { SupportMultipleContent = true })
            {
                try
                {
                    while (jsonReader.Read())
                    {
                        var prog = client.JsonSerializer.DeserializeObject<T>(jsonReader);
                        if (prog == null) continue;

                        progress.Report(prog);
                    }
                }
                catch (ObjectDisposedException)
                {
                    // The subsequent call to reader.ReadLineAsync() after cancellation
                    // will fail because we disposed the stream. Just ignore here.
                }
            }
        }
    }
}
```

```
internal class JsonSerializer
{
    public T DeserializeObject<T>(JsonReader jsonReader)
    {
        var serializer = Newtonsoft.Json.JsonSerializer.Create(this._settings);
        serializer.CheckAdditionalContent = false;
        return serializer.Deserialize<T>(jsonReader);
    }
   ...
}
```

**Additional information:**

### PR DESCRIPTION
**Pull request by dgvives**
Fixes #236 
Fixes #375
Fixes #416
Fixes #434
Closes #343 
Closes #351 
Closes #394

It solves execution stop on `DockerClient.Containers.WaitContainerAsync` waiting for the task to finish without being able to cancel.

```csharp
var waitContainerTask = dockerClient.Containers.WaitContainerAsync(createContainerResponse.ID, cts.Token);
```

- Task can be cancelled using token or default timeout.
- Included tests to verify cancellation works both ways.
- Included tests to several endpoints calling this method.
- Fixed tests failing on different execution contexts due to new exceptions being thrown.

### CODE REVIEW

**Comment by dgvives** [comment]
Working solution pending tests cases.

**Comment by dgvives** [comment]
@galvesribeiro @jterry75 and other maintainers, 

I'm developing integration tests for [IContainerOperations](https://github.com/dotnet/Docker.DotNet/blob/8e3cf4f7ac386e2f76791afc2c427a7b7884cd5e/src/Docker.DotNet/Endpoints/IContainerOperations.cs#L10) endpoint, since I need to test this piece of code via `IContainerOperations` endpoint. 

It would be better to just develop tests for affected functionality, right?
And include tests for currently untested endpoint methods on a different changeset, to keep it as small as possibletests

**Comment by dgvives** [comment]
Closed, tests produce different results on each execution

**Comment by dgvives** [comment]
Updated tests to include timeout

**Comment by dgvives** [comment]
Included tests to validate issues solution.

**Comment by dgvives** [comment]
Core changes are set to a minimum. However, some issues are complicated to test.
New test cases have been added to validate issue solution, and a common test fixture put in place to avoid collisions between test monitoring events.
All core modifications have been tested and tests are included within this PR.

@jterry75 @galvesribeiro would you have a look on it?

File: src/Docker.DotNet/Endpoints/StreamUtil.cs
Comment (jterry75) [inline src/Docker.DotNet/Endpoints/StreamUtil.cs:26]: Fascinating idea. You seem to have an extra `await` though.

File: src/Docker.DotNet/Endpoints/StreamUtil.cs
Comment (jterry75) [inline src/Docker.DotNet/Endpoints/StreamUtil.cs:18]: Why not in a using like the others?

**Review by jterry75 — COMMENTED** [review]
Just a couple of questions. Looks good

File: src/Docker.DotNet/Endpoints/StreamUtil.cs
Comment (dgvives) [reply src/Docker.DotNet/Endpoints/StreamUtil.cs:18]: Should be like others by next commit. I'm running tests now before pushing again

File: src/Docker.DotNet/Endpoints/StreamUtil.cs
Comment (dgvives) [reply src/Docker.DotNet/Endpoints/StreamUtil.cs:26]: Yes, it looks like that.

`await Task.WhenAny(task1, task2)` would return either task1 or task2.
Then, this task1 or task2 need to be awaited too.

Original idea from #343 by @suneption, I just adjusted it for tests to pass.

Comment (jterry75) [reply src/Docker.DotNet/Endpoints/StreamUtil.cs:26]: Crazy. I did not know you could inline await :).

File: src/Docker.DotNet/Endpoints/StreamUtil.cs
Comment (jterry75) [inline src/Docker.DotNet/Endpoints/StreamUtil.cs:25]: Alright one last question. When `tcs.Task` is selected the 2nd `await` will throw because there is no result. I assume this is the correct behavior and we want to throw the cancellation exception?

**Review by jterry75 — APPROVED** [review]
LGTM

**Comment by jterry75** [comment]
@galvesribeiro - Will you check my last question about throwing the cancellation exception here. As long as your good with that we can take this change IMO.

File: src/Docker.DotNet/Endpoints/StreamUtil.cs
Comment (dgvives) [reply src/Docker.DotNet/Endpoints/StreamUtil.cs:25]: When cancelling a token I expect to get an OpertionCancelledException. Otherwise it would mean task reached its end instead of being cancelled. 

I have no knowledge about architecture or system designs, i just enjoy thinks working on the way I like. Do the minimum to not do it twice.

I've included several tests on this, expecting OperationCancelledException to be thrown and failing test otherwise. 
There should be on IContainerOperationsTests

Comment (galvesribeiro) [reply src/Docker.DotNet/Endpoints/StreamUtil.cs:25]: I don't think it will throw there. The `tcs.Task` will be returned.  

In this case I wouldn't assign to the `line` variable directly. I would assign it to a Task and would check which task is that - the readline or the cancellation. 

If that is the cancellation task, then yeah, it is cancelled. If it is the deadline Task, then you `string line = await readLineTask();`.

Comment (dgvives) [reply src/Docker.DotNet/Endpoints/StreamUtil.cs:25]: There should be a test behind this

Comment (galvesribeiro) [reply src/Docker.DotNet/Endpoints/StreamUtil.cs:25]: Agreed.

**Comment by dgvives** [comment]
@jterry75 @galvesribeiro  to not merge this. Tested behavior is not right on `MonitorStreamAsync`

`MonitorStreamForMessagesAsync`, `MonitorResponseForMessagesAsync`, and `MonitorStreamAsync` behave in the same way, throwing **TaskCancelledException** when task is cancelled. 

Expect unified behavior to be pushed soon

**Comment by dgvives** [comment]
image

Some tests have output details, I used them to understand what I was doing. At some point I gave up but test output details remain

File: test/Docker.DotNet.Tests/IContainerOperationsTests.cs
Comment (Emdot) [inline test/Docker.DotNet.Tests/IContainerOperationsTests.cs:131]: This may suggest a deadlock bug in the code, rather than an xUnit problem. It's common for async code to work in Console apps but break elsewhere, due to the difference in synchronization contexts. Unfortunately, I'm not an expert in how sync contexts and schedulers and such work, so I don't know where to suggest looking.

File: src/Docker.DotNet/DockerClient.cs
Comment (Emdot) [inline src/Docker.DotNet/DockerClient.cs:355]: Would it make sense to move the timeout complexity to a separate function? Then this whole method can be:
```
var request = PrepareRequest(method, path, queryString, headers, data);
return RunWithTimeoutAsync(
    ct => _client.SendAsync(request, completionOption, ct), 
    cancellationToken, timeout);
```

A RunWithTimeoutAsync implementation might look like this:
```
public static async Task<T> TimeoutAfter<T>(Func<CancellationToken, Task<T>> makeTask, CancellationToken inputToken, TimeSpan timeout)
{
    var cts = new CancellationTokenSource();

    var timerTask = Task.Delay(timeout, cts.Token);
    var mainTask = makeTask(cts.Token);
    
    // Start both the timeout and the main task in parallel.
    Task firstTask == await Task.WhenAny(timerTask, mainTask).ConfigureAwait(false);
    
    // At this point, one of them has finished. Cancel the other one.
    cts.Cancel();
    
    // If the timeout occurred first, throw a timed-out exception.
    if (firstTask == timerTask)
        throw new TimeoutException();
    
    // Otherwise return the result of the main task. If it was canceled, this may throw 
    // a cancelation exception. If it faulted, it may throw some other exception.
    return await mainTask.ConfigureAwait(false);
}
```

For added performance you could use a state parameter to avoid the closure. As a bonus, timeouts emit a separate exception than cancellation.

Comment (dgvives) [reply src/Docker.DotNet/DockerClient.cs:355]: @Emdot great suggestion. 
I'll wait for this to get approved and continue with improvements.

File: test/Docker.DotNet.Tests/IContainerOperationsTests.cs
Comment (dgvives) [reply test/Docker.DotNet.Tests/IContainerOperationsTests.cs:131]: I see, I'm creating some sort of web app to test this. Will update with test results

Comment (dgvives) [reply test/Docker.DotNet.Tests/IContainerOperationsTests.cs]: My bad, failing test wasn't **IContainerOperations.GetContainerLogs_Tty_True_StreamReadsLogs** but **IContainerOpreations.GetContainerStatsAsync_StreamStats**

I didn't notice it because it works fine on build agent.

Comment (dgvives) [reply test/Docker.DotNet.Tests/IContainerOperationsTests.cs]: JsonReader.ReadAsync can't be cancelled once serialization has started.
Looking for an alternative now

**Comment by dgvives** [comment]
@jterry75 moving back to stream.Dispose() solved cancellation problems while monitoring streams. 
Tests have been added to validate failing tasks under different circumstances

**Comment by dgvives** [comment]
Closed while research on stream.read not being cancelled is done

File: src/Docker.DotNet/Endpoints/StreamUtil.cs
Comment (dgvives) [reply src/Docker.DotNet/Endpoints/StreamUtil.cs]: I listened to your advice. It's not exactly as you said but shares the same idea.
It helped fixing some problems when reading streams from containers created without a tty
Thank you @galvesribeiro

**Comment by dgvives** [comment]
When using `JsonSerializer.Deserialize<T>(JsonReader)` to deserialize, deep inside it is calling `ExpandoObject.ReadReadObject(JsonReader)` and no cancellation is possible from inside the loop, there is no token to cancel. 


I went over the source for JsonReader, StreamReader, Stream, and finally JsonSerializer until I found out this.


https://github.com/JamesNK/Newtonsoft.Json/blob/b6dc05be5a0f4808f06ec430f3bb59b24d3fbc3e/Src/Newtonsoft.Json/Converters/ExpandoObjectConverter.cs#L113#L141
```csharp
        private object ReadObject(JsonReader reader)
        {
            IDictionary<string, object?> expandoObject = new ExpandoObject();

            while (reader.Read())
            {
                switch (reader.TokenType)
                {
                    case JsonToken.PropertyName:
                        string propertyName = reader.Value!.ToString();

                        if (!reader.Read())
                        {
                            throw JsonSerializationException.Create(reader, "Unexpected end when reading ExpandoObject.");
                        }

                        object? v = ReadValue(reader);

                        expandoObject[propertyName] = v;
                        break;
                    case JsonToken.Comment:
                        break;
                    case JsonToken.EndObject:
                        return expandoObject;
                }
            }

            throw JsonSerializationException.Create(reader, "Unexpected end when reading ExpandoObject.");
        }
```



This hasn't been to easy to find out, difficult to figure out a solution, and turned even more complicated to test.

**Comment by dgvives** [comment]
@jterry75 apologies as I've been up and down with this PR. It turned out to be way more complicated than it seemed
@galvesribeiro thanks for your suggestion, please take a look at how it ended up

I'm having some rest after this

**Comment by jterry75** [comment]
@dgvives - Looks like you got it!

**Comment by dgvives** [comment]
@jterry75 - Yes, it has been a very interesting journey into task cancelling and tests

**Comment by jterry75** [comment]
Do you want to rebase all your commits into one just because it was a lot of change over them and then I'll do a final review and signoff. But looks good to me

**Comment by dgvives** [comment]
Done

**Comment by dgvives** [comment]
@jterry75 done

**Review by jterry75 — APPROVED** [review]
LGTM

**Comment by jterry75** [comment]
@galvesribeiro - Want to take a look as well

**Review by galvesribeiro — APPROVED** [review]
Awesome work! Thank you for the contribution and the huge patience across the reviews.

**Comment by galvesribeiro** [comment]
@jterry75 do you think we have enough for a new release now?

**Comment by galvesribeiro** [comment]
@jterry75 ?

**Pull request merged**
